File: app/src/main/java/uz/kidzone/app/AdsManager.java
package uz.kidzone.app;

import android.app.Activity;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;

import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.MobileAds;
import com.google.android.gms.ads.RequestConfiguration;
import com.google.android.gms.ads.interstitial.InterstitialAd;
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback;
import com.google.android.gms.ads.rewarded.RewardedAd;
import com.google.android.gms.ads.rewarded.RewardedAdLoadCallback;

/**
 * Implementation of IAdsManager using Google Mobile Ads (AdMob).
 * Adheres to SRP by focusing only on Advertisement lifecycle.
 */
public class AdsManager implements IAdsManager {
    private static final String TAG = "AdsManager";
    private static final int BANNER_RETRY_DELAY_MS = 30000;

    private final Activity activity;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    
    private BannerListener bannerListener;
    private AdView bannerAdView;
    private InterstitialAd interstitialAd;
    private RewardedAd rewardedAd;

    public AdsManager(Activity activity) {
        this.activity = activity;
    }

    @Override
    public void initialize() {
        configureAdMob();
        MobileAds.initialize(activity, status -> {
            Log.d(TAG, "AdMob Initialized");
            loadInterstitial();
            loadRewarded();
        });
    }

    private void configureAdMob() {
        RequestConfiguration config = new RequestConfiguration.Builder()
                .setTagForChildDirectedTreatment(RequestConfiguration.TAG_FOR_CHILD_DIRECTED_TREATMENT_TRUE)
                .setMaxAdContentRating(RequestConfiguration.MAX_AD_CONTENT_RATING_G)
                .build();
        MobileAds.setRequestConfiguration(config);
    }

    @Override
    public void loadBanner(ViewGroup container, boolean isTablet) {
        loadBanner(container, isTablet, null);
    }

    public void loadBanner(ViewGroup container, boolean isTablet, BannerListener listener) {
        cleanupBanner(container);
        this.bannerListener = listener;
        
        bannerAdView = new AdView(activity);
        AdSize adSize = isTablet ? AdSize.LEADERBOARD : AdSize.BANNER;
        bannerAdView.setAdSize(adSize);
        bannerAdView.setAdUnitId(activity.getString(R.string.banner_ad_unit_id));

        bannerAdView.setAdListener(new AdListener() {
            @Override
            public void onAdLoaded() {
                container.setVisibility(View.VISIBLE);
                if (bannerListener != null) bannerListener.onBannerLoaded(adSize.getHeight());
            }

            @Override
            public void onAdFailedToLoad(@NonNull LoadAdError e) {
                Log.e(TAG, "Banner failed: " + e.getMessage());
                if (bannerListener != null) bannerListener.onBannerFailed();
                mainHandler.postDelayed(() -> loadBanner(container, isTablet), BANNER_RETRY_DELAY_MS);
            }
        });

        container.addView(bannerAdView);
        bannerAdView.loadAd(createAdRequest());
    }

    private void cleanupBanner(ViewGroup container) {
        if (bannerAdView != null) {
            container.removeView(bannerAdView);
            bannerAdView.destroy();
            bannerAdView = null;
        }
    }

    @Override
    public void showInterstitial() {
        if (interstitialAd != null) {
            interstitialAd.show(activity);
            interstitialAd = null;
            loadInterstitial();
        } else {
            Log.d(TAG, "Interstitial not ready yet");
            loadInterstitial();
        }
    }

    private void loadInterstitial() {
        InterstitialAd.load(activity, activity.getString(R.string.interstitial_ad_unit_id),
                createAdRequest(), new InterstitialAdLoadCallback() {
                    @Override
                    public void onAdLoaded(@NonNull InterstitialAd ad) {
                        interstitialAd = ad;
                    }

                    @Override
                    public void onAdFailedToLoad(@NonNull LoadAdError loadAdError) {
                        interstitialAd = null;
                    }
                });
    }

    @Override
    public void showRewarded(OnRewardListener listener) {
        if (rewardedAd != null) {
            rewardedAd.show(activity, rewardItem -> listener.onReward(rewardItem.getAmount()));
            rewardedAd = null;
            loadRewarded();
        } else {
            Log.d(TAG, "Rewarded ad not ready");
            loadRewarded();
        }
    }

    private void loadRewarded() {
        RewardedAd.load(activity, activity.getString(R.string.rewarded_ad_unit_id),
                createAdRequest(), new RewardedAdLoadCallback() {
                    @Override
                    public void onAdLoaded(@NonNull RewardedAd ad) {
                        rewardedAd = ad;
                    }

                    @Override
                    public void onAdFailedToLoad(@NonNull LoadAdError loadAdError) {
                        rewardedAd = null;
                    }
                });
    }

    private AdRequest createAdRequest() {
        return new AdRequest.Builder().build();
    }

    @Override
    public void onResume() {
        if (bannerAdView != null) bannerAdView.resume();
    }

    @Override
    public void onPause() {
        if (bannerAdView != null) bannerAdView.pause();
    }

    @Override
    public void onDestroy() {
        if (bannerAdView != null) bannerAdView.destroy();
    }
}

File: app/src/main/java/uz/kidzone/app/AdsManager.kt
package uz.kidzone.app

import android.app.Activity
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.View
import android.view.ViewGroup
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.MobileAds
import com.google.android.gms.ads.RequestConfiguration
import com.google.android.gms.ads.interstitial.InterstitialAd
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback
import com.google.android.gms.ads.rewarded.RewardedAd
import com.google.android.gms.ads.rewarded.RewardedAdLoadCallback

/**
 * Implementation of IAdsManager using Google Mobile Ads (AdMob).
 * Adheres to SRP by focusing only on Advertisement lifecycle.
 */
class AdsManager(private val activity: Activity) : IAdsManager {

    companion object {
        private const val TAG = "AdsManager"
        private const val BANNER_RETRY_DELAY_MS = 30_000L
    }

    private val mainHandler = Handler(Looper.getMainLooper())

    private var bannerListener: IAdsManager.BannerListener? = null
    private var bannerAdView: AdView? = null
    private var interstitialAd: InterstitialAd? = null
    private var rewardedAd: RewardedAd? = null

    override fun initialize() {
        configureAdMob()
        MobileAds.initialize(activity) {
            Log.d(TAG, "AdMob Initialized")
            loadInterstitial()
            loadRewarded()
        }
    }

    private fun configureAdMob() {
        val config = RequestConfiguration.Builder()
            // COPPA compliance: Tag for child-directed treatment
            .setTagForChildDirectedTreatment(RequestConfiguration.TAG_FOR_CHILD_DIRECTED_TREATMENT_TRUE)
            // Google Play Families Policy: Max ad content rating (G = General)
            .setMaxAdContentRating(RequestConfiguration.MAX_AD_CONTENT_RATING_G)
            // GDPR/Underage: Tag for under age of consent
            .setTagForUnderAgeOfConsent(RequestConfiguration.TAG_FOR_UNDER_AGE_OF_CONSENT_TRUE)
            .build()
        MobileAds.setRequestConfiguration(config)
    }

    override fun loadBanner(container: ViewGroup, isTablet: Boolean) {
        loadBanner(container, isTablet, null)
    }

    fun loadBanner(container: ViewGroup, isTablet: Boolean, listener: IAdsManager.BannerListener?) {
        cleanupBanner(container)
        bannerListener = listener

        val adSize = if (isTablet) AdSize.LEADERBOARD else AdSize.BANNER
        bannerAdView = AdView(activity).also { adView ->
            adView.setAdSize(adSize)
            adView.adUnitId = activity.getString(R.string.banner_ad_unit_id)
            adView.adListener = object : AdListener() {
                override fun onAdLoaded() {
                    container.visibility = View.VISIBLE
                    bannerListener?.onBannerLoaded(adSize.height)
                }

                override fun onAdFailedToLoad(e: LoadAdError) {
                    Log.e(TAG, "Banner failed: ${e.message}")
                    bannerListener?.onBannerFailed()
                    mainHandler.postDelayed(
                        { loadBanner(container, isTablet) },
                        BANNER_RETRY_DELAY_MS
                    )
                }
            }
            container.addView(adView)
            adView.loadAd(createAdRequest())
        }
    }

    private fun cleanupBanner(container: ViewGroup) {
        bannerAdView?.let {
            container.removeView(it)
            it.destroy()
            bannerAdView = null
        }
    }

    override fun showInterstitial() {
        if (interstitialAd != null) {
            interstitialAd!!.show(activity)
            interstitialAd = null
            loadInterstitial()
        } else {
            Log.d(TAG, "Interstitial not ready yet")
            loadInterstitial()
        }
    }

    private fun loadInterstitial() {
        InterstitialAd.load(
            activity,
            activity.getString(R.string.interstitial_ad_unit_id),
            createAdRequest(),
            object : InterstitialAdLoadCallback() {
                override fun onAdLoaded(ad: InterstitialAd) {
                    interstitialAd = ad
                }

                override fun onAdFailedToLoad(e: LoadAdError) {
                    interstitialAd = null
                }
            }
        )
    }

    override fun showRewarded(listener: IAdsManager.OnRewardListener) {
        if (rewardedAd != null) {
            rewardedAd!!.show(activity) { rewardItem -> listener.onReward(rewardItem.amount) }
            rewardedAd = null
            loadRewarded()
        } else {
            Log.d(TAG, "Rewarded ad not ready")
            loadRewarded()
        }
    }

    private fun loadRewarded() {
        RewardedAd.load(
            activity,
            activity.getString(R.string.rewarded_ad_unit_id),
            createAdRequest(),
            object : RewardedAdLoadCallback() {
                override fun onAdLoaded(ad: RewardedAd) {
                    rewardedAd = ad
                }

                override fun onAdFailedToLoad(e: LoadAdError) {
                    rewardedAd = null
                }
            }
        )
    }

    private fun createAdRequest(): AdRequest {
        // Force Non-Personalized Ads (NPA) for all requests in kids' app
        val extras = Bundle().apply { putString("npa", "1") }
        return AdRequest.Builder()
            .addNetworkExtrasBundle(com.google.ads.mediation.admob.AdMobAdapter::class.java, extras)
            .build()
    }

    override fun onResume() {
        bannerAdView?.resume()
    }

    override fun onPause() {
        bannerAdView?.pause()
    }

    override fun onDestroy() {
        bannerAdView?.destroy()
    }
}
